harness/determined/debug.py

```diff
@@ -0,0 +1,137 @@
+import contextlib
+import faulthandler
+import threading
+import time
+from typing import Any, Callable, Dict, Generator
+
+import psutil
+
+
+@contextlib.contextmanager
+def stack_trace_thread(stack_trace_period_sec: float) -> Generator:
+    """If enabled, emit stack traces periodically."""
+    if stack_trace_period_sec <= 0.0:
+        yield
+        return
+
+    faulthandler.dump_traceback_later(stack_trace_period_sec, repeat=True)
+    try:
+        yield
+    finally:
+        faulthandler.cancel_dump_traceback_later()
+
+
+class ProfilingThread(threading.Thread):
+    """
+    Call a profiling function periodically on a separate thread (asynchronous collection).
+
+    The _ProfilingThread also supports synchronous triggering, which is useful for collecting
+    precise metrics around workload transitions.
+    """
+
+    def __init__(self, period: float, profile_fn: Callable) -> None:
+        self.period = period
+        self.profile_fn = profile_fn
+        self.quitting = False
+        self.cond = threading.Condition()
+        super().__init__()
+
+    def run(self) -> None:
+        """Asynchronously call a profiling function to collect metrics in a thread."""
+        with self.cond:
+            while True:
+                self.cond.wait(timeout=self.period)
+                if self.quitting:
+                    break
+                self.profile_fn()
+
+    def __enter__(self) -> None:
+        self.start()
+
+    def __exit__(self, *arg: Any) -> None:
+        self.quit()
+
+    @contextlib.contextmanager
+    def pause(self) -> Generator:
+        """Guarantee that the thread will not trigger asynchronously."""
+        self.cond.acquire()
+        try:
+            yield
+        finally:
+            self.cond.release()
+
+    def trigger(self) -> None:
+        """
+        Allow the profile_fn() to be triggered externally.  Only safe to call inside the context
+        manager provided by this self object.
+        """
+        self.profile_fn()
+
+    def quit(self) -> None:
+        self.quitting = True
+        with self.cond:
+            self.cond.notify_all()
+
+
+class Profiler:
+    """
+    Stateful metric collection. It is nice to have as a separate object from the ProfilingLayer for
+    flexibility in local training situations.
+    """
+
+    def __init__(self, system_level_metrics: bool, process_level_metrics: bool) -> None:
+        self.system_level_metrics = system_level_metrics
+        self.process_level_metrics = process_level_metrics
+
+        self.last_time = time.time()
+
+        # Bootstrap some stateful profiling calls.
+        if self.system_level_metrics:
+            _ = psutil.cpu_percent(percpu=True)
+
+            d = psutil.disk_io_counters()
+            self.sys_disk_read_bytes = d.read_bytes
+            self.sys_disk_write_bytes = d.write_bytes
+
+            n = psutil.net_io_counters()
+            self.sys_net_bytes_recv = n.bytes_recv
```

Comment:
We should be a little careful here in how we present this to the user - the network throughput numbers are misleading if they are sampled too infrequently and they won't correctly tell you if you're hitting the max your interconnect is capable of.

Reply:
What do you mean?  Like if the ints overflow?

Reply:
Network usage tends to look something like this:
![2_second_sample](https://user-images.githubusercontent.com/13806811/95922128-77584c00-0d67-11eb-9afa-128aeb236cd8.png)
If we sample infrequently, we could end up calculating an average throughput where it looks like you aren't bottlenecked by the network, but you actually are, we are just including periods with no network activity in the average. Depending on how we present this info to the user, this can confuse users when they are trying to figure out why distributed performance isn't good - https://github.com/horovod/horovod/issues/255#issuecomment-385570706

Reply:
Hm, this is a good point.  So we have a few of options here I think.
 - We could tell users "if you want to detect network bottlenecks you should set your profiling period very short"
 - We could have the resource profiling for different resources operate on different periods.
 - We could be more opinionated about how profiling should be done in general, and sample things like network usage at a much higher rate than things like CPU (which `psutil` says can't be meaningfully sample more than once every 0.1 seconds)

I think I'm in favor of the first option, with the understanding that at some point for this to be useful with very high resolution we'll have to add some buffering in there somewhere.

But given this information, I think I'll remove the `0.1` lower limit on resource period that I put in the code.

@armandmcqueen wdyt?

Reply:
Yeah I like option 1. 

It's probably not a major issue, but part of me would like to see the performance impact of frequent sampling either tested or mention that very short profiling periods could impact performance. I'm fine if we don't do that as part of this PR

Reply:
> very short profiling periods might impact training performance

Sounds good to me.  I'll have this in the docs.